lib/ansible/modules/clustering/k8s.py

```diff
@@ -0,0 +1,111 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2018, Chris Houseknecht <@chouseknecht>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+
+module: k8s
+
+short_description: Manage Kubernetes (K8s) objects
+
+version_added: "2.5"
+
+author: "Chris Houseknecht (@chouseknecht)"
+
+description:
+  - Use the OpenShift Python client to perform CRUD operations on Kubernetes objects.
+  - Supports authentication using either a config file, certificates, password or token.
+
+extends_documentation_fragment:
+  - kubernetes
+
+requirements:
+    - "python >= 2.7"
+    - "openshift >= 0.3"
+    - "PyYAML >= 3.11"
+'''
+
+EXAMPLES = '''
+- name: Create a k8s namespace
+  k8s:
+    name: testing
+    api_version: v1
+    kind: namespace
+    state: present
+
+- name: Create a Service object from an inline definition
+  k8s:
+    state: present
+    definition:
+      apiVersion: v1
+      kind: Service
+      metadata:
+        name: web
+        namespace: testing
+        labels:
+          app: galaxy
+          service: web
+      spec:
+        selector:
+          app: galaxy
+          service: web
+        ports:
+        - protocol: TCP
+          targetPort: 8000
+          name: port-8000-tcp
+          port: 8000
+
+- name: Create a Service object by reading the definition from a file
+  k8s:
+    state: present
+    src: /testing/service.yml
+
+- name: Remove an existing Service object
+  k8s:
+    state: absent
+    api_version: v1
+    kind: service
```

Comment:
If you pipe this through to the URL, then this should really be "resource".  In Kube "kind" is the schema, "resource" is the thing you access via a client.  Kind is always a constant (uppercase), resource is always the lowercase RESTful name.

I would recommend `resource: service` (if you're doing discovery) or `resource: services` (if you're hardcoding).  Mapping `kind` to `resource` is something we explicitly do on the client by looking up a Kind in the discovery doc, so overlapping them can be dangerous.

Reply:
@smarterclayton 

The `kind` module parameter is one-for-one with the `kind` option found in a Kubernetes config file. In both cases the value of `kind` is a model name.

Normally, in a config file, the model name is in uppercase. We're allowing, however, for the fact that the user will likely provide a lowercased value, and thus the lowercase `service` or `namespace` value.

I updated the doc strings to reflect exactly what the `kind` parameter is: https://github.com/chouseknecht/ansible/blob/feature/k8s_module/lib/ansible/utils/module_docs_fragments/kubernetes.py#L63

Hopefully that clarifies things, and lines up with your thinking.